File: app/src/main/java/com/prey/activities/PasswordHtmlActivity.java
/*******************************************************************************
 * Created by Orlando Aliaga
 * Copyright 2020 Prey Inc. All rights reserved.
 * License: GPLv3
 * Full license at "/LICENSE"
 ******************************************************************************/
package com.prey.activities;

import android.app.Activity;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.view.KeyEvent;
import android.view.View;
import android.view.WindowManager;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.window.OnBackInvokedCallback;
import android.window.OnBackInvokedDispatcher;

import com.prey.PreyConfig;
import com.prey.PreyLogger;
import com.prey.PreyUtils;
import com.prey.R;
import com.prey.activities.js.CustomWebView;
import com.prey.activities.js.WebAppInterface;

public class PasswordHtmlActivity extends Activity {

    private WebView myWebView = null;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        // Fullscreen immersive + show over lock screen
        getWindow().addFlags(
            WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON
            | WindowManager.LayoutParams.FLAG_DISMISS_KEYGUARD
            | WindowManager.LayoutParams.FLAG_SHOW_WHEN_LOCKED
            | WindowManager.LayoutParams.FLAG_TURN_SCREEN_ON
        );

        setContentView(R.layout.webview);

        // Block back button/gesture (Android 13+ predictive back)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            getOnBackInvokedDispatcher().registerOnBackInvokedCallback(
                OnBackInvokedDispatcher.PRIORITY_DEFAULT,
                new OnBackInvokedCallback() {
                    @Override
                    public void onBackInvoked() {
                        // Do nothing — block back while locked
                    }
                }
            );
        }

        // Re-hide navigation bar whenever it appears
        getWindow().getDecorView().setOnSystemUiVisibilityChangeListener(new View.OnSystemUiVisibilityChangeListener() {
            @Override
            public void onSystemUiVisibilityChange(int visibility) {
                if ((visibility & View.SYSTEM_UI_FLAG_FULLSCREEN) == 0) {
                    hideSystemUI();
                }
            }
        });
        PreyLogger.d("PasswordHtmlActivity: onCreate");

        myWebView = (WebView) findViewById(R.id.install_browser);
        myWebView.setOnKeyListener(new View.OnKeyListener() {
            @Override
            public boolean onKey(View view, int i, KeyEvent keyEvent) {
                CustomWebView.callDispatchKeyEvent(getApplicationContext(),keyEvent);
                return false;
            }
        });
        WebSettings settings = myWebView.getSettings();
        myWebView.setBackgroundColor(0x00000000);
        settings.setJavaScriptEnabled(true);
        settings.setLoadWithOverviewMode(true);
        settings.setLoadsImagesAutomatically(true);
        settings.setUseWideViewPort(true);
        settings.setSupportZoom(false);
        settings.setBuiltInZoomControls(false);
        String lng = PreyUtils.getLanguage();
        String url = "%s#/%s/%s";
        String lockMessage = PreyConfig.getPreyConfig(this).getLockMessage();
        if (lockMessage != null && !"".equals(lockMessage)) {
            url = String.format(url, CheckPasswordHtmlActivity.URL_ONB, lng, "lockmessage");
        } else {
            url = String.format(url, CheckPasswordHtmlActivity.URL_ONB, lng, "lock");
        }
        myWebView.addJavascriptInterface(new WebAppInterface(this, this), CheckPasswordHtmlActivity.JS_ALIAS);
        myWebView.loadUrl(url);
        PreyConfig.getPreyConfig(this).viewLock = myWebView;
    }

    @Override
    protected void onResume() {
        super.onResume();
        hideSystemUI();

        String unlockPass = PreyConfig.getPreyConfig(getApplicationContext()).getUnlockPass();
        boolean isLock = unlockPass != null && !"".equals(unlockPass);
        PreyLogger.d("PasswordHtmlActivity isLock:" + isLock);
        if (!isLock) {
            finishAffinity();
        }
    }

    @Override
    public void onWindowFocusChanged(boolean hasFocus) {
        super.onWindowFocusChanged(hasFocus);
        if (hasFocus) {
            hideSystemUI();
        }
    }

    private void hideSystemUI() {
        getWindow().getDecorView().setSystemUiVisibility(
            View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY
            | View.SYSTEM_UI_FLAG_FULLSCREEN
            | View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
            | View.SYSTEM_UI_FLAG_LAYOUT_STABLE
            | View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
            | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
        );
    }

    @Override
    public void onBackPressed() {
        // Block back button while locked
    }

    public void pfinish() {
        finishAffinity();
    }
}

File: app/src/main/java/com/prey/activities/PasswordHtmlActivity.kt
/*******************************************************************************
 * Created by Orlando Aliaga
 * Copyright 2025 Prey Inc. All rights reserved.
 * License: GPLv3
 * Full license at "/LICENSE"
 ******************************************************************************/
package com.prey.activities

import android.app.Activity
import android.os.Bundle
import android.view.View
import android.webkit.WebView

import com.prey.R
import com.prey.activities.js.CustomWebView
import com.prey.activities.js.WebAppInterface
import com.prey.PreyConfig
import com.prey.PreyLogger
import com.prey.PreyUtils

/**
 * Activity responsible for displaying the password HTML page.
 */
class PasswordHtmlActivity : Activity() {
    private var myWebView: WebView? = null

    /**
     * Called when the activity is created.
     *
     * @param savedInstanceState Saved instance state, if any.
     */
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.webview)
        PreyLogger.d("PasswordHtmlActivity: onCreate")
        val unlockPass: String? = PreyConfig.getInstance(applicationContext).getUnlockPass()
        val isLock = unlockPass != null && "" != unlockPass
        PreyLogger.d("PasswordHtmlActivity isLock:$isLock")
        if (!isLock) {
            PreyConfig.getInstance(this).setLoadUrl(URL_OUT)
            finishAffinity()
            return
        }
        myWebView = findViewById<View>(R.id.install_browser) as WebView
        myWebView!!.setOnKeyListener { view, i, keyEvent ->
            CustomWebView.callDispatchKeyEvent(applicationContext, keyEvent)
            false
        }
        val settings = myWebView!!.settings
        myWebView!!.setBackgroundColor(0x00000000)
        settings.javaScriptEnabled = true
        settings.loadWithOverviewMode = true
        settings.loadsImagesAutomatically = true
        settings.useWideViewPort = true
        settings.setSupportZoom(false)
        settings.builtInZoomControls = false
        val lng: String = PreyUtils.getLanguage()
        var url = "${CheckPasswordHtmlActivity.URL_ONB}#/${lng}/"
        val lockMessage: String? = PreyConfig.getInstance(applicationContext).getLockMessage()
        if (lockMessage != null && "" != lockMessage) {
            url += URL_LOCK_WIH_MESSAGE
        } else {
            url += URL_LOCK
        }
        myWebView!!.addJavascriptInterface(
            WebAppInterface(this, this),
            CheckPasswordHtmlActivity.JS_ALIAS
        )
        PreyConfig.getInstance(applicationContext).setLoadUrl(url)
        myWebView!!.loadUrl(url)
        PreyConfig.getInstance(this).setLoadUrl(url)
        myWebView!!.loadUrl("javascript:window.location.reload(true)")
        PreyConfig.getInstance(this).viewLock = myWebView
    }

    companion object {
        const val URL_LOCK: String = "lock"
        const val URL_LOCK_WIH_MESSAGE: String = "lockmessage"
        const val URL_OUT: String = "out"
    }

}